src/viem/prepareTransactions.ts

```diff
@@ -0,0 +1,206 @@
+    maxGasCost += BigInt(tx.gas) * BigInt(tx.maxFeePerGas)
+  }
+  return new BigNumber(maxGasCost.toString())
+}
+
+function getFeeCurrencyAddress(feeCurrency: TokenBalance) {
+  return !feeCurrency.isNative ? (feeCurrency.address as Address) : undefined
+}
+
+async function tryEstimateTransaction({
+  baseTransaction,
+  maxFeePerGas,
+  feeCurrencyAddress,
+  maxPriorityFeePerGas,
+}: {
+  baseTransaction: TransactionRequestCIP42
+  maxFeePerGas: bigint
+  feeCurrencyAddress?: Address
+  maxPriorityFeePerGas?: bigint
+}) {
+  const tx = {
+    ...baseTransaction,
+    maxFeePerGas,
+    maxPriorityFeePerGas,
+    // Don't include the feeCurrency field if not present.
+    // See https://github.com/wagmi-dev/viem/blob/e0149711da5894ac5f0719414b4ecc06ccaecb7b/src/chains/celo/serializers.ts#L164-L168
+    ...(feeCurrencyAddress && { feeCurrency: feeCurrencyAddress }),
+  }
+
+  // TODO maybe cache this? and add static padding when using non-native fee currency
+  try {
+    tx.gas = await publicClient.celo.estimateGas({
+      ...tx,
+      account: tx.from,
+    })
+  } catch (e) {
+    if (e instanceof EstimateGasExecutionError) {
+      // Likely too much gas was needed
+      Logger.warn(
+        'SwapScreen@useSwapQuote',
+        `Couldn't estimate gas with feeCurrencyAddress ${feeCurrencyAddress}`,
+        e
+      )
+      return null
+    }
+    throw e
+  }
+
+  return tx
+}
+
+async function tryEstimateTransactions(
+  baseTransactions: TransactionRequestCIP42[],
+  feeCurrency: TokenBalance
+) {
+  const transactions: TransactionRequestCIP42[] = []
+
+  const feeCurrencyAddress = getFeeCurrencyAddress(feeCurrency)
+  const { maxFeePerGas, maxPriorityFeePerGas } = await estimateFeesPerGas(
+    publicClient.celo,
+    feeCurrencyAddress
+  )
+
+  for (const [index, baseTx] of baseTransactions.entries()) {
+    // We can only truly estimate the gas for the first transaction
+    if (index === 0) {
+      const tx = await tryEstimateTransaction({
+        baseTransaction: baseTx,
+        feeCurrencyAddress,
+        maxFeePerGas,
+        maxPriorityFeePerGas,
+      })
+      if (!tx) {
+        return null
+      }
+      transactions.push(tx)
+    } else {
+      if (!baseTx.gas) {
+        throw new Error(
+          'When multiple transactions are provided, all transactions but the first one must have a gas value'
+        )
+      }
+      transactions.push({
+        ...baseTx,
+        maxFeePerGas,
+        maxPriorityFeePerGas,
+        // Don't include the feeCurrency field if not present.
+        // See https://github.com/wagmi-dev/viem/blob/e0149711da5894ac5f0719414b4ecc06ccaecb7b/src/chains/celo/serializers.ts#L164-L168
+        ...(feeCurrencyAddress && { feeCurrency: feeCurrencyAddress }),
+        // We assume the provided gas value is with the native fee currency
+        // If it's not, we add the static padding
+        gas: baseTx.gas + BigInt(feeCurrency.isNative ? 0 : STATIC_GAS_PADDING),
+      })
+    }
+  }
```

Comment:
Like you said in https://github.com/valora-inc/wallet/pull/4368#discussion_r1372395257
> seems like there are cases this might not be true, like if we wanted to use this for estimating gas for registering the DEK and doing a c-stable transfer

I proposed to add an option to control this https://github.com/valora-inc/wallet/pull/4368#discussion_r1372726574
> true, maybe this could be an option we pass? { isDependantTransactions: boolean }

But maybe we could do that later as an optimization? if we really want/need this.

Reply:
thanks. in my efforts to keep my PR in sync with yours I ended up copying over things that I questioned on your PR lol. I updated it